src/irc/core/irc-cap.c

```diff
@@ -0,0 +1,96 @@
+#include "module.h"
+#include "signals.h"
+#include "misc.h"
+
+#include "irc-cap.h"
+#include "irc-servers.h"
+
+static void event_cap (IRC_SERVER_REC *server, char *args, char *nick, char *address)
+{
+	GSList *tmp;
+	GString *req_cmd;
+	char *params, *evt, *list;
+	char **caps, *sig_name;
+	int i;
+
+	params = event_get_params(args, 3, NULL, &evt, &list);
+	caps = g_strsplit(list, " ", -1);
+
+	if (!strcmp(evt, "LS")) {
+		/* create a list of the supported caps */
+		for (i = 0; caps[i] != NULL; i++)
+			server->cap = g_slist_append(server->cap, g_strdup(caps[i]));
+
+		/* request the required caps, if any */
+		if (!server->cap_req) {
+			server->cap_complete = TRUE;
+			irc_send_cmd_now(server, "CAP END");
+		}
+
+		req_cmd = g_string_new("CAP REQ :");
+
+		for (tmp = server->cap_req; tmp != NULL; tmp = tmp->next) {
+			/* check whether the cap is supported by the server */ 
+			if (!gslist_find_string(server->cap, tmp->data)) {
+				server->cap_complete = TRUE;
+				irc_send_cmd_now(server, "CAP END");
+
+				g_string_free(req_cmd, TRUE);
+				return;
+			}
+
+			g_string_append_printf(req_cmd, "%s ", tmp->data);
+		}
+
+		irc_send_cmd_now(server, req_cmd->str);
+
+		g_string_free(req_cmd, TRUE);
+	}
+	else if (!strcmp(evt, "ACK")) {
+		int got_sasl = FALSE;
+
+		/* emit a signal for every ack'd cap */
+		for (i = 0; caps[i] != NULL; i++) {
+			if (!strcmp(caps[i], "sasl"))
+				got_sasl = TRUE;
+
+			sig_name = g_strdup_printf("server cap %s", caps[i]);
+			signal_emit(sig_name, 1, server);
+			g_free(sig_name);
+		}
+
+		/* the negotiation is now over, unless 'sasl' is requested.
+		 * in that case the negotiation will be closed after the login succeeds */
+		if (got_sasl == FALSE) {
```

Comment:
Not really sure if it's okay to do this here. Maybe there could be other ircv3 extensions that require delaying registration? Maybe could pass a pointer to a variable like ``signal_emit(sig_name, 2, &continue_registration)` and the handler of the "server cap sasl" would set that to true?

*(removed old post and reposted in the right place)*